File: src/Edmunds/Gae/RuntimeEnvironment.php
<?php

/**
 * Edmunds
 *
 * The fast PHP framework for building web applications.
 *
 * @license   This file is subject to the terms and conditions defined in file 'license.md', which is part of this source code package.
 */

namespace Edmunds\Gae;

use Edmunds\Patterns\Singleton;

/**
 * Environment check for Google App Engine
 */
class RuntimeEnvironment
{
	use Singleton;

	/**
	 * 'true' if running on GAE.
	 *
	 * @var bool
	 */
	protected $gaeEnvironment;

	/**
	 * Constructor
	 */
	public function __construct()
	{
		$this->detectGae();
	}

	/**
	 * Detect if the application is running on GAE.
	 */
	protected function detectGae()
	{
		$appIdentityService = 'google\appengine\api\app_identity\AppIdentityService';

		if ( ! class_exists($appIdentityService))
		{
			$this->gaeEnvironment = false;
			return;
		}

		$this->gaeEnvironment = (bool) $appIdentityService::getDefaultVersionHostname();
	}

	/**
	 * Returns 'true' if running on GAE.
	 *
	 * @return bool
	 */
	public function isGae()
	{
		return $this->gaeEnvironment;
	}
}

File: edmunds/gae/runtimeenvironment.py

import os
import re


class RuntimeEnvironment(object):
    """
    The Google App Engine runtime environment
    """

    @staticmethod
    def is_gae():
        """
        Check if is Google App Engine
        """

        if 'CURRENT_VERSION_ID' not in os.environ:
            return False

        if 'AUTH_DOMAIN' not in os.environ:
            return False

        if 'SERVER_SOFTWARE' not in os.environ:
            return False

        server_software = os.environ.get('SERVER_SOFTWARE')

        if not re.match(r'^Development\/', server_software) and not re.match(r'^Google App Engine\/', server_software):
            return False

        return True

    @staticmethod
    def is_gae_development():
        """
        Check if Google App Engine SDK
        :return:    Gae Development
        :rtype:     bool
        """
        if not RuntimeEnvironment.is_gae():
            return False

        return 'SERVER_SOFTWARE' in os.environ \
               and os.environ['SERVER_SOFTWARE'].startswith('Development')
